src/zulip.rs

```diff
@@ -303,48 +305,94 @@ struct MembersApiResponse {
 struct Member {
     email: String,
-    user_id: i64,
+    user_id: u64,
+}
+
+#[derive(serde::Serialize)]
+#[serde(tag = "type")]
+#[serde(rename_all = "snake_case")]
+pub enum Recipient<'a> {
+    Stream {
+        #[serde(rename = "to")]
+        id: u64,
+        topic: &'a str,
+    },
+    Private {
+        #[serde(skip)]
+        id: u64,
+        #[serde(rename = "to")]
+        email: &'a str,
+    },
+}
+
+impl Recipient<'_> {
+    pub fn narrow(&self) -> String {
+        match self {
+            Recipient::Stream { id, topic } => {
+                // See
+                // https://github.com/zulip/zulip/blob/46247623fc279/zerver/lib/url_encoding.py#L9
+                // ALWAYS_SAFE without `.` from
+                // https://github.com/python/cpython/blob/113e2b0a07c/Lib/urllib/parse.py#L772-L775
+                //
+                // ALWAYS_SAFE doesn't contain `.` because Zulip actually encodes them to be able
+                // to use `.` instead of `%` in the encoded strings
+                const ALWAYS_SAFE: &str =
+                    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_-~";
+
+                let mut encoded_topic = Vec::new();
+                for ch in topic.bytes() {
+                    if !(ALWAYS_SAFE.contains(ch as char)) {
+                        write!(encoded_topic, "%{:02X}", ch).unwrap();
+                    } else {
+                        encoded_topic.push(ch);
+                    }
+                }
+
+                let mut encoded_topic = String::from_utf8(encoded_topic).unwrap();
+                encoded_topic = encoded_topic.replace("%", ".");
+
+                format!("stream/{}-xxx/topic/{}", id, encoded_topic)
+            }
+            Recipient::Private { id, .. } => format!("pm-with/{}-xxx", id),
+        }
+    }
 }
 
 #[derive(serde::Serialize)]
 pub struct MessageApiRequest<'a> {
-    #[serde(rename = "type")]
-    pub type_: &'a str,
-    pub to: &'a str,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub topic: Option<&'a str>,
+    pub recipient: Recipient<'a>,
     pub content: &'a str,
 }
 
-impl MessageApiRequest<'_> {
-    // FIXME: support private links too
-    pub fn url(&self) -> String {
-        // See
-        // https://github.com/zulip/zulip/blob/46247623fc279/zerver/lib/url_encoding.py#L9
-        // ALWAYS_SAFE from
-        // https://github.com/python/cpython/blob/113e2b0a07c/Lib/urllib/parse.py#L772-L775
-        let always_safe = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_.-~";
-        let mut encoded_topic = Vec::new();
-        for ch in self.topic.expect("topic").bytes() {
-            if !(always_safe.contains(ch as char) || ch == b'.') {
-                write!(encoded_topic, "%{:02X}", ch).unwrap();
-            } else {
-                encoded_topic.push(ch);
-            }
+impl<'a> MessageApiRequest<'a> {
+    pub fn private(user_id: u64, user_email: &'a str, content: &'a str) -> Self {
```

Comment:
I am not a fan of this as it has two &str fields without much protection for getting it right, and the u64 is also not obviously just one thing. I would prefer that we construct the ApiRequest directly at the call sites for both of these constructor functions.